product/cmd/grpc/client/client_test.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"context"
+	"net"
+	"testing"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+
+	"product/internal/grpc/domain"
+	"product/internal/grpc/services"
+	repo "product/internal/repository"
+)
+
+type FakeProductRepository struct {
+	store []*domain.Product // temporary storage for products
+}
+
+func (fp *FakeProductRepository) CreateProduct(p *domain.Product) (*domain.Product, error) {
+	// temporary store product
+	fp.store = append(fp.store, p)
+	return p, nil
+}
+
+var fakeRepo = &FakeProductRepository{}
+
+type productService struct{}
+
+func (s productService) AddProduct(ctx context.Context, in *domain.AddProductRequest) (*domain.AddProductResponse, error) {
+	return &domain.AddProductResponse{
+		CreatedProduct: fakeRepo.store[0],
+	}, nil
+}
+
+// duck-typing at work
+// just making sure the fake implements the ProductRepository interface
+var fakeStore repo.ProductRepository = fakeRepo
```

Comment:
You can do this without having both `fakeStore` and `fakeRepo`:

`var fakeStore repo.ProductRepository = &FakeProductRepository{}`

If I want to guarantee that something stays in line with an interface then I sometimes add:

`var _ InterfaceToAlignWith = (*MyType)(nil)`